Implement redirect bridge to support COOP (#8118)
AzureAD/microsoft-authentication-library-for-js: lib/msal-browser/docs/v4-migration.md

1. The `protocolMode` parameter has been moved to `SystemOptions` from `BrowserAuthOptions` in Configuration. There are no changes to its options or functionality.
1. The `navigateFrameWait` parameter has been removed. This was previously needed by older browsers which are no longer supported by MSAL.js.

> 1. The `iframeHashTimeout` and `windowHashTimeout` parameters have been replaced with `iframeBridgeTimeout` and `popupBridgeTimeout` respectively. These timeouts now control how long to wait for a response from the redirect bridge via BroadcastChannel API when COOP is enabled.

Reviewer: ```suggestion
1. The `iframeHashTimeout` and `windowHashTimeout` parameters have been replaced with `iframeBridgeTimeout` and `popupBridgeTimeout` respectively. These timeouts now control how long to wait for a response from the redirect bridge via BroadcastChannel API.
```

Isn't this used now regardless of COOP?

> Note: In cases where MSAL determines `extraParameters` must be encoded into the URL string, `extraParameters` will be merged with `extraQueryParams` in a way that will cause same-named parameters to be overwritten. In these cases, the value for the parameter in `extraParameters` will take precedence over the value in the `extraQueryParams`.

### Cross-Origin-Opener-Policy (COOP) Support

> MSAL Browser v5 introduces built-in support for Cross-Origin-Opener-Policy (COOP), which enhances security by isolating your browsing context. When COOP is enabled on your application, popup and iframe-based authentication flows are restricted. MSAL v5 provides a redirect bridge mechanism to handle authentication in COOP-enabled environments.

Reviewer: Isn't COOP going to be enabled server side? Does application policy matter anymore once enabled by the service?

Author: Fixed

#### What Changed

When your application has COOP headers enabled (e.g., `Cross-Origin-Opener-Policy: same-origin`), traditional popup and silent iframe authentication flows will fail because the authentication window cannot communicate back to the main application window. MSAL v5 solves this by introducing a redirect bridge pattern.

> **All authentication flows** (`acquireTokenSilent()`, `ssoSilent()`, `loginPopup()`, and `loginRedirect()`) now use the redirect bridge when COOP is enabled. The redirect bridge handles the authentication response differently based on the flow:

Reviewer: Always right?
```suggestion
**All authentication flows** (`acquireTokenSilent()`, `ssoSilent()`, `loginPopup()`, and `loginRedirect()`) now use the redirect bridge. The redirect bridge handles the authentication response differently based on the flow:
```

- **Popup and silent flows**: The redirect bridge broadcasts the authentication response to the main application window using the BroadcastChannel API

> - **Redirect flow**: The redirect bridge navigates back to your application's home page with the authentication response in the URL

Reviewer: ```suggestion
- **Redirect flow**: The redirect bridge navigates back to your application's page that initiated the redirect with the authentication response in the URL
```

#### How It Works

> 1. **Main application**: Your COOP-enabled application initiates authentication using `loginPopup()`, `ssoSilent()`, or `loginRedirect()`

Reviewer: ```suggestion
1. **Main application**: Your application initiates authentication using `loginPopup()`, `ssoSilent()`, or `loginRedirect()`
```

        broadcastResponseToMainFrame()
            .catch((error) => {
                console.error("Error broadcasting response:", error);
            });
    </script>
</body>
</html>
```

##### 2. Configure your web server

Ensure your redirect URI page is served **without** COOP headers, while your main application pages have COOP enabled:

```javascript
// Example using Express.js
app.get("/redirect", (req, res) => {
    // DO NOT set COOP headers for redirect page
    res.sendFile(__dirname + "/redirect.html");
});

app.get("*", (req, res) => {
    // Set COOP headers for all other pages
    res.setHeader("Cross-Origin-Opener-Policy", "same-origin");
    res.sendFile(__dirname + "/index.html");
});
```

##### 3. Update your MSAL configuration

Set the `redirectUri` in your MSAL configuration to point to your redirect bridge page:

```javascript
const msalConfig = {
    auth: {
        clientId: "{your-client-id}",
        authority: "https://login.microsoftonline.com/common",
        redirectUri: "https://{your-app-home-page}/redirect", // Point to your redirect bridge page
    },
};

const pca = new PublicClientApplication(msalConfig);
```

> - For a complete working example of COOP support with the redirect bridge, see the [COOP Sample Application](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-browser-samples/COOP).

Reviewer: Let's take this out since the sample is temporary and for our own E2E use. I don't want to forget to remove the reference here.
```suggestion
```

**Note:** Microsoft Entra ID (formerly Azure AD) has COOP enabled by default. For Azure AD B2C, COOP availability depends on your backend configuration and the authentication endpoints being used.

#### What Changed

When COOP headers are present on the authentication service response (e.g., `Cross-Origin-Opener-Policy: same-origin`), traditional popup and silent iframe authentication flows will fail because the authentication window cannot communicate back to the main application window. MSAL v5 solves this by introducing a redirect bridge pattern.

**All authentication flows** (`acquireTokenSilent()`, `ssoSilent()`, `loginPopup()`, and `loginRedirect()`) now use the redirect bridge when COOP is enabled. The redirect bridge handles the authentication response differently based on the flow:

- **Popup and silent flows**: The redirect bridge broadcasts the authentication response to the main application window using the BroadcastChannel API

> - **Redirect flow**: The redirect bridge navigates back to your application's home page with the authentication response in the URL

Reviewer 2: This remains unchanged. Should we mention it?

Author: Can you elaborate please? All flows have changed. The redirect bridge is always invoked.

#### How It Works

> 1. **Main application**: Your COOP-enabled application initiates authentication using `loginPopup()`, `ssoSilent()`, or `loginRedirect()`

Reviewer 2: Again non blocking comment: Visual helps. We can use mermaid which is super fast to generate visuals and can checkin the source code.

Author: Added visuals to express and react-router samples. Links and additional docs are located in `login-user.md`

2. **Redirect**: MSAL opens a popup/iframe/window to an authority page
3. **Authentication flow**: The authority page completes the OAuth flow and receives the auth response
4. **Response handling**: The redirect page uses the new `broadcastResponseToMainFrame()` function which:
    - For **popup/silent flows**: Broadcasts the response to the main window via BroadcastChannel API

>     - For **redirect flows**: Navigates to your application's home page with the auth response

Reviewer 2: Navigate to `page where the `acquireTokenRedirect` is initiated from` correct?

Author: Updated doc
